Canny destination doc (#807)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/canny.mdx

1. From your [RudderStack dashboard](https://app.rudderstack.com/), add a source. Then, from the list of destinations, select **Canny**.
2. Assign a name to the destination and click on **Continue**.

## Connection settings

To successfully configure Canny as a destination, you will need to configure the following settings:

<img src="../../assets/event-stream-destinations/canny-connection-settings.png" alt="Canny connection settings"/>

- **API Key**: Enter your API key from the Canny dashboard.

<div class="infoBlock">
Refer to the <Link to="#faq">FAQ</Link> section below for more information on finding the API Key.
</div>

### Event settings

- **Mapping to trigger Canny Events for the respective Event**: Enter the event name and pick the Canny event from the dropdown to be triggered when that event is called.

<div class="successBlock">
You can specify multiple <strong>Canny Events</strong> for one <strong>Event Name</strong> and vice versa.
</div>

## Identify

You can use the <Link to="/event-spec/standard-events/identify/">`identify`</Link> call to find the `id` of a user:
- If a user does not exist, it is created and its `id` is returned. 

> - If a user exists, it is updated with new data.

Reviewer: We can create and update a user only through identify call.

|:---------------------|:-------------|:--------------|:--------------|
| `properties.boardID` | `boardID` | Required | String |
| Collect from `externalId` or <Link to="https://developers.canny.io/api-reference#retrieve_user">retrieve user</Link> API call| `authorID` | Required | String |
| `properties.details` | `details` | Required | String |
| `properties.title` | `title` | Required | String |
| `properties.byID` | `byID` | Optional | String |
| `properties.categoryID` | `categoryID` |Optional | String |
| `properties` (after removing `boardID`, `byID`, `categoryID`,`details`, `eta`, `etaPublic`, `title`, `imageURLs`)| `customFields` | Optional | Object |
| `properties.eta` | `eta` | Optional | String |
| `properties.etaPublic` | `etaPublic` | Optional | Boolean |
| `properties.imageURLs` | `imageURLs` | Optional | Array of strings |

### Property mappings for create vote Event 

The following table lists the mappings between RudderStack and Canny properties for **Create vote** event:

| RudderStack property | Canny property | Presence | Data type | 
|:---------------------|:-------------|:--------------|:--------------|
| `properties.postId` | `postID` | Required | String |

> | Collect from `externalId` or <Link to="https://developers.canny.io/api-reference#retrieve_user">retrieve user</Link> API call | `voterID` | Required | String |

Reviewer: The key name of externalId is `cannyUserId` we have to mention that somewhere.

A sample `track` call is shown below:

```javascript
rudderanalytics.track("Submit", {
    boardId: "62de88676bc28b44eeef25dd",
    details: "Require new feature",
    title: "New Feature",
    eta: "10.08.2022",
    priority: "High"
})
```

### Property mappings for Create post event 

The following table lists the mappings between RudderStack and Canny properties for **Create post** event:

| RudderStack property | Canny property | Presence | Data type | Notes |
|:---------------------|:-------------|:--------------|:--------------|:--------------|
| `properties.boardID`/`properties.boardId`/`properties.board.id` | `boardID` | Required | String | -|
| Collect from `externalId` or <Link to="https://developers.canny.io/api-reference#retrieve_user">retrieve user</Link> API call| `authorID` | Required | String |-|
| `properties.details` | `details` | Required | String |-|
| `properties.title` | `title` | Required | String |-|
| `properties.byID`/`properties.byId`/`properties.by.id` | `byID` | Optional | String |-|
| `properties.categoryID`/`properties.categoryId`/`properties.category.id` | `categoryID` |Optional | String |-|

> | `properties` (after removing `boardID`, `byID`, `categoryID`,`details`, `eta`, `etaPublic`, `title`, `imageURLs`)| `customFields` | Optional | Object | Ensure that you <Link to="#how-to-create-custom-fields-in-canny-dashboard">create the custom fields in Canny dashboard</Link> to send the event seuccessfully.|

Reviewer: for `customFields` we are looking for `customFields` in properties so need to update the mapping.
```suggestion
| `properties.customFields` | `customFields` | Optional | Object | Ensure that you <Link to="#how-to-create-custom-fields-in-canny-dashboard">create the custom fields in Canny dashboard</Link> to send the event seuccessfully.|
```

| Connection Mode | Web | Mobile | Server |
| :--- | :--- | :--- | :--- |
| Device mode | - | - | - |
| Cloud mode |Supported  | Supported | Supported |

<div class="infoBlock">
To know more about the difference between cloud mode and device mode in RudderStack, refer to the <Link to="/destinations/rudderstack-connection-modes/">RudderStack Connection Modes</Link> guide.
</div>

Once you have confirmed that the source platform supports sending events to Canny, follow these steps:

1. From your [RudderStack dashboard](https://app.rudderstack.com/), add a source. Then, from the list of destinations, select **Canny**.
2. Assign a name to the destination and click on **Continue**.

## Connection settings

To successfully configure Canny as a destination, you will need to configure the following settings:

<img src="../../assets/event-stream-destinations/canny-connection-settings.png" alt="Canny connection settings"/>

- **API Key**: Enter your API key from the Canny dashboard.

<div class="infoBlock">
Refer to the <Link to="#faq">FAQ</Link> section below for more information on getting the API Key.
</div>

### Event settings

> - **Mapping to trigger Canny Events for the respective Event**: Enter the event name and pick the Canny event from the dropdown to be triggered when that event is called.

Reviewer 2: ```suggestion
- **Mapping to trigger Canny Events for the respective Event**: Enter the event name and choose the Canny event from the dropdown to be triggered when that event is called.
```

<div class="successBlock">
You can specify multiple <strong>Canny Events</strong> for one <strong>Event Name</strong> and vice versa.
</div>

## Identify

> You can use the <Link to="/event-spec/standard-events/identify/">`identify`</Link> call to <Link to="https://developers.canny.io/api-reference#create_or_update_user">create or update a user</Link>.

Reviewer 2: ```suggestion
You can use the <Link to="/event-spec/standard-events/identify/">`identify`</Link> call to <Link to="https://developers.canny.io/api-reference#create_or_update_user">create or update a user</Link> in Canny.
```

A sample `identify` call is shown below:

```javascript
rudderanalytics.identify("1hKOmRA4el9Zt1WSfVJIVo4GRlm", {
    "name": "Alex Keener",
    "email": "alex@example.com",
    "gender": "Male",
    "profession": "singer-songwriter",
    "companies": [{
        "created": "2020-01-23T04:56:07.890Z",
        "customFields": {
            "field1": "value1",
        },
        "id": "company123",
        "monthlySpend": 500.00,
        "name": "company name"
    }]
});
```

### Property mappings

The following table lists the mappings between RudderStack and Canny properties:

| RudderStack property | Canny property | Presence | Data type | 
|:---------------------|:-------------|:--------------|:--------------|
| `userId`/`traits.userId`/`traits.id`/`context.traits.userId`/`context.traits.id` | `userId` | Required | String |
| `context.traits.name`/`traits.name` | `name` |Required | String |
| `context.traits.email`/`traits.email`/`properties.email` | `email` | Optional | String |
| `context.traits`/`traits` (after removing `email` and `name`) | `customFields` |Optional | Object |

> | `originalTimestamp`/`timestamp` | `created` | Optional | Date |

Reviewer 2: Should this be in the ISO 8601 timestamp format? @ItsSudip @rohithkaza cc @ishachopra0401 we should specify this accordingly.

Reviewer: Yes, it should be in the ISO 8601 timestamp format.

> | `originalTimestamp`/`timestamp` | `created` | Optional | Date |

Reviewer 2: ```suggestion
| `originalTimestamp`/`timestamp` | `created` | Optional | ISO 8601 Timestamp |
```

| `context.traits.avatarURL`/`traits.avatarURL` | `avatarURL` | Optional | String |
| `context.traits.companies`/`traits.companies` | `companies` | Optional | Array of object (in company objects `id` and `name` are the only required fields) |

## Track

You can create the following events using the <Link to="/event-spec/standard-events/track/">`track`</Link> call:

- <Link to="https://developers.canny.io/api-reference#create_post">Create post</Link>
- <Link to="https://developers.canny.io/api-reference#create_vote">Create vote</Link>
 

> Also, you can retrieve the user data using `userId` or `email` and use that information to post or vote.

Reviewer 3: ```suggestion
We <Link to="/event-spec/standard-events/track/">`rerieve the user`</Link> data using `userId` or `email` and use that information to create post or vote.
```

A sample `track` call is shown below:

```javascript
rudderanalytics.track("Submit", {
    boardId: "62de88676bc28b44eeef25dd",
    details: "Require new feature",
    title: "New Feature",
    eta: "10.08.2022",
    priority: "High"
})
```

### Property mappings for Create post event 

> The following table lists the mappings between RudderStack and Canny properties for **Create post** event:

Reviewer 2: ```suggestion
The following table lists the mappings between RudderStack and Canny properties for the **Create post** event:
```

| RudderStack property | Canny property | Presence | Data type | Notes |

> |:---------------------|:-------------|:--------------|:--------------|:--------------|

Reviewer 2: we can drop the **Notes** column here.

| `properties.boardID`/`properties.boardId`/`properties.board.id` | `boardID` | Required | String | -|
| Collect from `externalId` or <Link to="https://developers.canny.io/api-reference#retrieve_user">retrieve user</Link> API call| `authorID` | Required | String |-|
| `properties.details` | `details` | Required | String |-|
| `properties.title` | `title` | Required | String |-|
| `properties.byID`/`properties.byId`/`properties.by.id` | `byID` | Optional | String |-|
| `properties.categoryID`/`properties.categoryId`/`properties.category.id` | `categoryID` |Optional | String |-|

> | `properties.customFields` | `customFields` | Optional | Object | Ensure that you <Link to="#how-to-create-custom-fields-in-canny-dashboard">create the custom fields in Canny dashboard</Link> to send the event seuccessfully.|

Reviewer 2: ```suggestion
| `properties.customFields` | `customFields` | Optional | Object | Ensure that you <Link to="#how-to-create-custom-fields-in-canny-dashboard">create the custom fields in Canny dashboard</Link> to send the event successfully.|
```

> | `properties.eta` | `eta` | Optional | String | In MM/YYYY format. Required if `etaPublic` is present. |

Reviewer 2: ```suggestion
| `properties.eta` | `eta` | Required if `etaPublic` is present. | String | Should be in MM/YYYY format.  |
```

> | `properties.etaPublic` | `etaPublic` | Optional | Boolean | Required if `eta` is present. |

Reviewer 2: ```suggestion
| `properties.etaPublic` | `etaPublic` | Required if `eta` is present.  | Boolean | - |
```

| `properties.imageURLs` | `imageURLs` | Optional | Array of strings | -|

### Property mappings for Create vote event 

> The following table lists the mappings between RudderStack and Canny properties for **Create vote** event:

Reviewer 2: ```suggestion
The following table lists the mappings between RudderStack and Canny properties for the **Create vote** event:
```

| RudderStack property | Canny property | Presence | Data type | 
|:---------------------|:-------------|:--------------|:--------------|
| `properties.postId`/`properties.postID`/`properties.post.id` | `postID` | Required | String |
| Collect from `externalId` or <Link to="https://developers.canny.io/api-reference#retrieve_user">retrieve user</Link> API call | `voterID` | Required | String |

<div class="infoBlock">
The key name for <code class="inline-code">externalId</code> is <code class="inline-code">cannyUserId</code>.
</div>

## FAQ

### Where can I find the Canny API Key?

To find the Canny API Key:

1. Log into your [Canny account](https://canny.io/login).
2. Go to **Settings** > **API & Webhooks** under your profile, as shown: 

<img src="../../assets/event-stream-destinations/canny-api-key.png" alt="Canny dashboard"/>

> ### How to create custom fields in Canny Dashboard?

Reviewer 2: ```suggestion
### How to create custom fields in the Canny dashboard?
```

To create custom fields in Canny Dashboard:

1. Log into your [Canny account](https://canny.io/login).
2. Go to **Settings** > **Boards** under your profile.
3. Select the desired **Board** from the dropdown and select **Create Post Form**.

> 4. Click on **Create new Post Fields to add them to a board** in the **Fields** section.

Reviewer 2: ```suggestion
4. In the **Fields** section, click on **Create new Post Fields to add them to a board**.
```

5. Click on **Create New Field**, as shown:

> 

Reviewer: We need to do another step:
Your Company/Dashboard -> Settings-> Boards -> [Select a Board From Dropdown] -> Create Post Form -> Add Field -> [Select a Post Field From Dropdown] -> Enter details -> Save
